src/Features/CSharp/Portable/CodeRefactorings/ConvertConversionOperators/ConvertConversionOperatorsRefactoringProvider.cs

```diff
@@ -0,0 +1,86 @@
+using Microsoft.CodeAnalysis.CodeRefactorings.ConvertConversionOperators;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.ConvertConversionOperators
+{
+    /// <inheritdoc/>
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = PredefinedCodeRefactoringProviderNames.ConvertConversionOperators), Shared]
+    internal partial class CSharpConvertConversionOperatorsRefactoringProvider : AbstractConvertConversionOperatorsRefactoringProvider<CastExpressionSyntax, BinaryExpressionSyntax>
+    {
+        [ImportingConstructor]
+        [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
+        public CSharpConvertConversionOperatorsRefactoringProvider()
+        {
+        }
+
+        protected override Task<ImmutableArray<BinaryExpressionSyntax>> FilterAsExpressionCandidatesAsync(ImmutableArray<BinaryExpressionSyntax> asExpression, AsyncLazy<SemanticModel> semanticModelFactory, CancellationToken cancellationToken)
+        {
+            var result = asExpression.WhereAsArray(binaryExpression =>
+                               binaryExpression.IsKind(SyntaxKind.AsExpression) &&
+                               binaryExpression.Right is TypeSyntax typeSyntax &&
+                               !typeSyntax.IsMissing);
+
+            return Task.FromResult(result);
+        }
+
+        protected override async Task<ImmutableArray<CastExpressionSyntax>> FilterCastExpressionCandidatesAsync(ImmutableArray<CastExpressionSyntax> castExpressions, AsyncLazy<SemanticModel> semanticModelFactory, CancellationToken cancellationToken)
+        {
+            if (castExpressions.IsEmpty)
+            {
+                return castExpressions;
+            }
+
+            var semanticModel = await semanticModelFactory.GetValueAsync(cancellationToken).ConfigureAwait(false);
+            var candidates = from node in castExpressions
+                             let type = semanticModel.GetTypeInfo(node, cancellationToken).Type
+                             where type != null && !type.IsValueType
+                             select node;
+
+            return candidates.ToImmutableArray();
+        }
+
+        protected override async Task<Document> ConvertFromAsToCastAsync(Document document, BinaryExpressionSyntax asExpression, CancellationToken cancellationToken)
+        {
+            var expression = asExpression.Left;
+            if (asExpression.Right is not TypeSyntax typeNode)
+            {
+                throw new InvalidOperationException("asExpression.Right must be a TypeSyntax. This check is done before the CodeAction registration.");
+            }
```

Comment:
make this a contract check, not an if.

Reply:
I do get CS0165 for `typeNode` (Use of uninitialized variable) later if I use `Contract.Fail` here. If I use 
`Contract.ThrowIfFalse(asExpression.Right is TypeSyntax)` I need to introduce the casted `typeNode` later again. Throwing the exception avoids both problems.